### ISSUE TITLE
Trying to re-open a deleted terminal gives you a new terminal

### ISSUE THREAD

**Issue by HarmtH** [neovim/neovim#5476]
- `nvim --version`: NVIM v0.1.6-234-g8cc760a                                                                             
- Vim (version: ) behaves differently? n/a
- Operating system/version: n/a
- Terminal name/version: n/a
- `$TERM`: n/a

### Actual behaviour
Re-opening a deleted terminal gives you a new terminal, with a different name.

### Expected behaviour
Throw an error.

### Steps to reproduce using `nvim -u NORC`

```
nvim -u NORC
:term
<c-\><c-n>
:argadd
:bd!
:argu
:argu
:argu
```

Now we have 3 new terminal buffers, because the buffer name in the argument list contains the PID, and terminal buffers with a different PID, i.e. different name, are created.

**Comment by HarmtH** [comment]
I think a good fix with be to check if a PID is given in the name passed to the ```term://``` BufReadCmd handler. If that is the case, do not create a terminal buffer, but instead throw an error.

**Comment by justinmk** [comment]
But it would break session restoration I think. Can the BufReadCmd somehow correct the args list, to replace the old name with a new name with a valid pid?

**Comment by HarmtH** [comment]
Didn't thought about session restoration. Also tried some BufReadCmd magic, but it seems you can't tell to Vim that you are unable to fulfil the request and want to error out (i.e. not create the buffer). The ```man://``` handler and netrw's handlers also don't do that.

I think the argument list entries can be fixed, but it has to be a ```noau tabdo windo``` construction then, as any window in any tab can have the terminal buffer in its argument list, and Vim doesn't have any construction to change another window's argument list.

**Comment by justinmk** [comment]
>  it seems you can't tell to Vim that you are unable to fulfil the request and want to error out (i.e. not create the buffer)

Why shouldn't it create the buffer? Here's what I was thinking: if `:argu` is called on an invalid `:term://pid`, 

1. BufReadCmd creates a new terminal buffer, 
2. finds the old buffer name in `argv()`, 
3. deletes the old entry from the args list, 
4. adds the new buffer name at the same args index.

> Vim doesn't have any function to change another window's argument list.

Gah, I didn't know args were window-local. But I'm not seeing that behavior: I've tried `:argadd` and `:argdelete` and they seem to affect all windows. `:args` reports the same list in each window and tabpage.

**Comment by justinmk** [comment]
I see `:arglocal` and the help says:

> There can be several argument lists.  They can be shared between windows.                                                                                   
> When they are shared, changing the argument list in one window will also                                                                                    
> change it in the other window.

If local argument lists are not connected, in each case the BufReadCmd can just follow the simple behavior described above. It will create a new buffer one time, and update whatever arglist happens to be updated by `:argdelete` and `:argadd`. The fact that some other local arglist may exist which still has an old terminal name with an invalid pid is not very important. The important thing is that the (local) arglist is updated and re-editing the entry won't re-create the terminal for that arglist.

What do you think?

**Comment by HarmtH** [comment]
> If local argument lists are not connected, in each case the BufReadCmd can just follow the simple behavior described above.

They originally pointed to the same terminal, but will point to different terminals, so I would say the approach of fixing each window's arglist at once would be a slightly better approach. I currently have the following in my vimrc:
```
    autocmd BufDelete *
                \ if expand('<afile>') =~ '^term://' |
                \   let curwin=winnr() |
                \   let curtab=tabpagenr() | 
                \   exe 'noau keepalt tabdo windo try | argd <afile> | catch | endtry' |
                \   keepalt noau exe 'normal' curtab.'gt'.curwin."\<c-w>\<c-w>" |
                \ endif
```
Which just deletes terminals from every arglist when they are closed. I don't notice any slowdown, even with very much windows open. Something alike could be put in BufReadCmd, but checking every argument in the arglist for a match would costs more resources of course.

**Comment by justinmk** [comment]
> They originally pointed to the same terminal, but will point to different terminals with that appraoch

Yes, why shouldn't they? User is editing an invalid terminal. I would recommend keeping it simple.

**Comment by HarmtH** [comment]
Ok, so in BufReadCmd:
- loop through the current arglist, delete the entries and mark the locations
- create a new term buffer
- add the new term buffer at the marked locations

**Comment by justinmk** [comment]
> loop through the current arglist, delete the entries and mark the locations

Shouldn't need to loop, `argidx()` will return the index that needs to be `:argdel`'d and re-added with the new buffer name.

If the old invalid term buffer name exists at multiple argument indexes, again, that's beyond the scope of what `BufReadCmd` should care about.

**Comment by HarmtH** [comment]
Fair point, placing this code after the current BufReadCmd seems to work and keeps the position of the termbuffer in the arglist the same:
```if expand('<amatch>')==argv(argidx()) | .argadd | .argd | endif```
